src/aio.rs

```diff
@@ -403,180 +394,168 @@ where
+        let this = self.as_mut().project();
+        this.sink_stream.poll_close(cx).map_err(|err| {
             self.send_result(Err(err));
         })
     }
 }
 
-impl<T> Pipeline<T>
+impl<SinkItem, I, E> Pipeline<SinkItem, I, E>
 where
-    T: Sink<SinkError = <T as Stream>::Error> + Stream + Send + 'static,
-    T::SinkItem: Send,
-    T::Item: Send,
-    T::Error: Send,
-    T::Error: ::std::fmt::Debug,
+    SinkItem: Send + 'static,
+    I: Send + 'static,
+    E: Send + 'static,
 {
-    fn new<E>(sink_stream: T, executor: E) -> Self
+    fn new<T, F>(sink_stream: T, mut executor: F) -> Self
     where
-        E: Executor<Box<dyn Future<Item = (), Error = ()> + Send>>,
+        F: Spawn,
+        T: Sink<SinkItem, Error = E> + Stream<Item = Result<I, E>> + 'static,
+        T: Send + 'static,
+        T::Item: Send,
+        T::Error: Send,
+        T::Error: ::std::fmt::Debug,
     {
         const BUFFER_SIZE: usize = 50;
         let (sender, receiver) = mpsc::channel(BUFFER_SIZE);
         let f = receiver
-            .map_err(|_| ())
-            .forward(PipelineSink {
-                sink_stream,
-                in_flight: VecDeque::new(),
-            })
+            .map(Ok)
+            .forward(PipelineSink::new::<SinkItem>(sink_stream))
             .map(|_| ());
-
-        executor.execute(Box::new(f)).unwrap();
+        executor.spawn(f).expect("Unable to spawn redis task");
         Pipeline(sender)
     }
 
     // `None` means that the stream was out of items causing that poll loop to shut down.
-    fn send(
-        &self,
-        item: T::SinkItem,
-    ) -> impl Future<Item = T::Item, Error = Option<T::Error>> + Send {
+    async fn send(&mut self, item: SinkItem) -> Result<I, Option<E>> {
         self.send_recv_multiple(item, 1)
-            .map(|mut item| item.pop().unwrap())
+            .map_ok(|mut item| item.pop().unwrap())
+            .await
     }
 
-    fn send_recv_multiple(
-        &self,
-        input: T::SinkItem,
+    async fn send_recv_multiple(
+        &mut self,
+        input: SinkItem,
         count: usize,
-    ) -> impl Future<Item = Vec<T::Item>, Error = Option<T::Error>> + Send {
-        let self_ = self.0.clone();
-
+    ) -> Result<Vec<I>, Option<E>> {
         let (sender, receiver) = oneshot::channel();
 
-        self_
+        self.0
             .send(PipelineMessage {
                 input,
                 response_count: count,
                 output: sender,
             })
             .map_err(|_| None)
             .and_then(|_| {
-                receiver.then(|result| match result {
-                    Ok(result) => result.map_err(Some),
-                    Err(_) => {
-                        // The `sender` was dropped which likely means that the stream part
-                        // failed for one reason or another
-                        Err(None)
-                    }
+                receiver.then(|result| {
```

Comment:
Instead of `then(|result| { future::ready(...) })`, I think you can simplify to just `.map()`?